### DISCUSSION TITLE
External files that require multiple extraction steps

### DISCUSSION THREAD

**Discussion by woolsweater** [twpayne/chezmoi#2682] [Q&A] [answered]
What would be the recommended way to handle external files that require more elaborate extraction than a single command?

As an example, I'd like make [Apple's SF fonts](https://developer.apple.com/fonts/) part of my chezmoi-mananged setup. They only distribute these on disk images, which contain compressed archives in a special Apple format. So the steps to install the font files would be something like:

1. Download the disk image
1. Mount it
1. Expand the archive on the image out to a temporary location
1. Copy the actual font files from the expanded directory tree into their destination (~/Library/Fonts/)
1. _Unmount_ the disk image
1. (Ideally but optionally) delete the download

It doesn't seem like this is the kind of process that `.chezmoiexternal.filter` is intended to support. Which is understandable, since it's rather more than a simple "filter". Is the preferred option therefore to simply use a `run_once` script to perform the required steps? This is fine but lacks that "declarative" sparkle, so I wanted to double-check. 🙂 

Chezmoi is a wonderful tool; thanks for your work!

**Comment by twpayne** [comment, marked as answer]
One simple solution is to add the extracted font files in ~/Library/Fonts directly to your dotfiles repo, i.e. run

```console
$ chezmoi add ~/Library/Fonts/SF-*.otf
```

This is declarative, at the expense of making updates to the fonts a manual process.

It would be theoretically possible to use `.chezmoiexternal.filter` where the filter is a script that you've written that performs the necessary steps to convert the `.dmg` into a an archive that chezmoi can unpack. However, this would be very inefficient in practice  because chezmoi would need to apply the filter every time you run `chezmoi diff` or `chezmoi apply` to see if anything has changed.

Although declarative approaches are nice, in practice a `run_once_` script is likely the best fit here.

**Comment by woolsweater** [comment]
That makes sense! I will stick with the `run_once` script; it does seem like the best option. Thanks!

**Reply by halostatue** [reply]
An alternative is to use `7z` (7zip or p7zip):

`7z e -r <DMG> '*.ttf' -o <target-path>`

**Reply by halostatue** [reply]
There are some 7z libraries for Go, but I suspect that they are for the 7z *format*, not the full features of 7z.

**Reply by twpayne** [reply]
Note that in this particular case the `.otf` files are in a `.pkg` installer inside the `.dmg`, so a single unpacking stage is not sufficient.

**Reply by halostatue** [reply]
True. `7z` can *probably* deal with this, too:

`7z e <DMG> <pkgfilename> -so | 7z e -r -si '*.ttf' -o <targetpath>`

This is *completely untested*, and is extremely fragile, obviously.

**Reply by woolsweater** [reply]
Interesting tip, thanks. I'll keep this in mind for the future.

**Reply by twpayne** [reply]
Wow, this actually worked. Using 7-Zip 22.01, it actually takes three stages of unpacking (!). I had to go via intermediate files as I think some of the stages require the ability to seek in the input, which you don't get with a pipe.

```console
$ 7zz e -so SF-Mono.dmg "SFMonoFonts/SF Mono Fonts.pkg" > "SF Mono Fonts.pkg"
$ 7zz e -so "SF Mono Fonts.pkg" Payload~ | tar xf - --strip-components=3
```

**Reply by halostatue** [reply]
`-so` outputs to stdout only, so the first line doesn’t need the redirect.

```console
$ 7zz e SF-Mono.dmg "SFMonoFonts/SF Mono Fonts.pkg" # writes "SF Mono Fonts.pkg" to the current directory
$ 7zz e -so "SF Mono Fonts.pkg" Payload~ | tar xf - --strip-components=3
```
